docs/proposals/decouple-application-sync-user-using-impersonation

```diff
@@ -0,0 +1,463 @@
+
+### Implementation Details/Notes/Constraints
+
+#### Component : GitOps Engine
+
+Fix GitOps Engine code to honor Impersonate configuration set in the Application sync context for all kubectl commands that are being executed.
+
+#### Component: ArgoCD API
+
+- Support new property `serviceAccountName` in `Destination` section of `AppProject`
+- Documentation of the new fields
+
+#### Component: ArgoCD Application Controller
+
+- Provide a configuration in argocd-cm  which can be modified to enable the Impersonation feature. Set `applicationcontroller.enable.impersonation: true` in the Argo CD ConfigMap.
+- Provide an option to override the Impersonation feature using environment variables.
+- Pass `--enable-impersonation` to the Application controller args.
+- Set `ARGOCD_APPLICATION_CONTROLLER_ENABLE_IMPERSONATION=true` in the Application controller environment variables.
+- Fix Application Controller sync.go to set the Impersonate configuration from the Application CR to the SyncContext Object (rawConfig and restConfig field, need to understand which config is used for the actual sync and if both configs need to be impersonated.)
+#### Component: ArgoCD UI
+
+- Provide option to provide `serviceAccountName` for a `Destination` section in an `AppProject` created/updated via the web console.
+- Update the User Guide documentation on how to use these newly added fields from the web console.
+#### Component: ArgoCD CLI
+
+- Provide option to provide serviceAccountName for an Application created/updated via the console.
+- Provide option to provide defaultServiceAccount for an AppProject created/updated via the web console.
+- Update the User Guide and other documentation where the CLI option usages are explained.
+
+
+### Detailed examples
+
+#### Example 1: Service account for application sync specified at the AppProject level for all namespaces
+In this specific scenario, service account name `generic-deployer` will get used for the application sync as the namespace `guestbook` matches the regex `*`.
+
+- Install ArgoCD in the `argocd` namespace.
+```
+kubectl apply -f https://raw.githubusercontent.com/argoproj/argo-cd/master/manifests/install.yaml -n argocd
+```
+- Enable the impersonation feature in ArgoCD.
+```
+kubectl set env statefulset/argocd-application-controller ARGOCD_APPLICATION_CONTROLLER_ENABLE_IMPERSONATION=true
+```
+- Create a namespace called `guestbook` and a service account called `guestbook-deployer`.
+```
+kubectl create namespace guestbook
+kubectl create serviceaccount guestbook-deployer
+```
+- Create Role and RoleBindings and configure RBAC access for creating `Service` and `Deployment` objects in namespace `guestbook` for service account `guestbook-deployer`.
+
+```
+kubectl create role guestbook-deployer-role --verb get,list,update,delete --resource pods,deployment,service
+kubectl create rolebinding guestbook-deployer-rb --serviceaccount guestbook-deployer --role guestbook-deployer-role
+```
```

Comment:
Can you add  example on how to allow or grant impersonation to `serviceaccount` `argocd` so it can impersonate `serviceaccount`  `guestbook-deployer` ?

the  `rolebinding` equivalent to:
```
oc create clusterrole impersonate-username --verb impersonate --resource users --resource-name username --as system:admin
oc adm policy add-cluster-role-to-user impersonate-username guestbook-deployer 
```

Reply:
Have provided an example where ArgoCD managed default service account `argocd-manager` would not be used.